py-development.Rmd

```diff
@@ -0,0 +1,540 @@
+them for more complicated tasks, which makes code more readable and easier to
+debug. By composing a function from the built-in `len()` and `sum()` functions,
+we can create a more succinct and easier to read version of our previous
+function `sum_and_len_all_numbers()`.
+
+```{python}
+def sum_and_len(list_of_numbers):
+    return sum(list_of_numbers), len(list_of_numbers)
+```
+
+#### Defining default values {#py-dev-defining-default-values}
+
+Up until now, our function calls have included enough arguments to assign one
+to each of the function parameters. This can get tedious in functions with many
+parameters and it is often beneficial to define default values for several of
+the function parameters. As an example, we can modify `sum_two_numbers()` to
+optionally return the two input arguments.
+
+```{python}
+def sum_two_numbers(num1, num2, return_input=False):
+    if return_input:
+        return num1, num2, num1 + num2
+    else:
+        return num1 + num2
+```
+
+By default the function will work just as previously.
+
+```{python}
+sum_two_numbers(2, 5)
+```
+
+But we now also have the option to return the input numbers.
+
+```{python}
+sum_two_numbers(2, 5, return_input=True)
+```
+
+Since the arguments are given in the same order as in the function definition,
+we could have left out the `return_input=` part and just written `True` in the
+third position. If we want to supply the arguments out of order, we need to
+specify the parameter name of each argument.
+
+#### Function documentation {#py-dev-function-documentation}
+
+Functions might appear self-explanatory when they are being written, but it is
+essential that there is proper documentation describing what the function does
+and what types of arguments should be in the input. This helps other people who
+are reading your code and also your future self that will be reusing these
+functions.
+
+In Python, a function is documented in its docstring, which is a multiline
+Python string immediately following the function definition. It is surrounded
+by triple quotes (either single or double) and can look like this.
+
+```{python}
+def sum_two_numbers(num1, num2):
+    '''
+    Add two numbers
+
+    Args:
+        num1 (int, float): The first number to be added.
+        num2 (int, float): The second number to be added.
+
+    Returns:
+        int, float: Sum of the two numbers.
+    '''
+
+    return num1 + num2
+```
+
+The above style is commonly referred to as the Google docstring convention. As
+functions get longer it can be beneficial to adapt the more elaborate `numpy`
+docstring convention used in many scientific Python packages:
+
+```{python}
+def sum_two_numbers(num1, num2):
+    '''
+    Add two numbers
+
+    Parameters
+    ----------
+    num1: int, float
+        The first number to be added.
+    num2: int, float
+        The second number to be added.
+
+    Returns
+    -------
+    int, float
+        Sum of the two numbers.
+    '''
```

Comment:
Does this render into a nice format or this is how it looks in the help? Does indentation matter here? Does parameter order matter?

Reply:
Good point to bring up. Spyder has rich rendering of docstrings treating it similarly (or maybe exactly as) markdown. This means the NumPy-like docstrings work slightly better as the arguments are line separated without adding extra trailing spaces or newlines in the source:

Numpy docstring format:
![image](https://user-images.githubusercontent.com/4560057/64229334-5b188900-ce9e-11e9-94f0-23987f1d6302.png)

Google dostring format:
![image](https://user-images.githubusercontent.com/4560057/64229574-075a6f80-ce9f-11e9-83a6-a168102a2845.png)

I also prefer numpy docstring and they are the most common among scientific packages, so I think I will make a harder recommendation for Numpy format, exclude the Google example and replace it with a link, and add a screenshot of the rendered docstring.

In PyCharm, I believe both renders as rich text, in jupyterlab the raw string is shown.

Reply:
I believe the docstrings are parsed the same as markdown, so yes indentation will matter since it makes the parameter name and description render as a markdown definition list with the parameter name in bold typeface. The order doesn't matter for anything that I am aware of but it is convention to keep the order the same as in the function signature.